### ISSUE TITLE
implement "incremental lists"

### ISSUE THREAD

**Issue by mhuesch** [sacredcapital/ReputationDSL#12]
[currently](https://github.com/mhuesch/poly-rs/blob/e21bfcbca3e3e1ddcffc940f09583bd665e85bea/src/eval.rs#L13):

```
#[derive(Clone)]
pub enum Value {
    ...
    VList(Vec<Value>),
    ...
}
```

change to

```
#[derive(Clone)]
pub enum Value {
    ...
    VList(Iterator<Item = Value>),
}
```

- [ ] figure out what places in the codebase clone `Value`
- [ ] figure out how to clone iterators (is it possible?)

**Comment by mhuesch** [comment]
as context:

if we clone the iterator "fully" (as in, the code inside of the iterator body is cloned) then we will duplicate effectful actions which might be enclosed in the iterator body. for example, we might read from Holochain to get the next value, or we might pull from the network 😱. so if we have 2 copies of the iterator, they would do these effectful actions twice. this could be a problem...

one way to mock-up "purity" (or rather, only singular execution of effects) is to have a cache which saves values once they are pulled from the iterator. thus, once we duplicate an iterator, it would only ever read from the cache, so the effect only occurs once.

in a recent meeting we decided that it's ok to duplicate effects for Holochain, so I dropped this. documenting it here in case we want to revisit in the future.
